core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,1255 @@
+    let interactionRulesService: InteractionRulesService;
+    let answerClassificationResult: AnswerClassificationResult = {
+      answerGroupIndex: 0,
+      classificationCategorization: 'default_outcome',
+      outcome: {
+        dest: null,
+        feedback: {
+          _contentId: 'feedback_id',
+          _html: '<p>Dummy Feedback</p>',
+          contentId: 'feedback_id',
+          html: '<p>Dummy Feedback</p>',
+        },
+        labelledAsCorrect: true,
+        missingPrerequisiteSkillId: null,
+        paramChanges: [],
+        refresherExplorationId: null,
+      },
+      ruleIndex: 0
+    } as unknown as AnswerClassificationResult;
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.returnValue(Promise.resolve(exploration));
+    let answerClassificationServiceSpy =
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
+        .and.returnValue(answerClassificationResult);
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => {
+        if (html === null) {
+          return null;
+        }
+        return html;
+      });
+
+    questionPlayerEngineService.init(
+      multipleQuestionBackendDict, successHandler, failHandler);
+    let currentQuestion1 = questionPlayerEngineService.getCurrentQuestion();
+    expect(currentQuestion1.getId()).toBe('questionId1');
+
+    questionPlayerEngineService.submitAnswer(
+      answer, interactionRulesService, successCallback);
+    questionPlayerEngineService.recordNewCardAdded();
+    let currentQuestion2 = questionPlayerEngineService.getCurrentQuestion();
+
+    expect(answerClassificationServiceSpy).toHaveBeenCalled();
+    expect(currentQuestion2.getId()).toBe('questionId2');
+  });
+
+  it('should return the current question Id', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+    let exploration = sampleExplorationDict;
+
+    spyOn(contextService, 'setQuestionPlayerIsOpen').and.returnValue(null);
+    spyOn(contextService, 'getExplorationId').and.returnValue('explorationId1');
+    spyOn(contextService, 'isInQuestionPlayerMode').and.returnValue(true);
+    spyOn(urlService, 'getExplorationVersionFromUrl').and.returnValue(1);
+    spyOn(readOnlyExplorationBackendApiService, 'loadExplorationAsync')
+      .and.returnValue(Promise.resolve(exploration));
+
+    questionPlayerEngineService.init(
+      multipleQuestionBackendDict, successHandler, failHandler);
+    let currentQuestionId = questionPlayerEngineService.getCurrentQuestionId();
+
+    expect(currentQuestionId).toBe('questionId1');
```

Comment:
Probably need a pre-condition test as well (before the init()) to make sure it's null and that the init() therefore changed it?

Reply:
We can not directly access the questionId ["getCurrentQuestionId()"] before init, since it is dependent on "this.questions" which will be initialized only after init.

Reply:
So what happens when getCurrentQuestionId() is called before init() (does it return null or throw an error)?

Reply:
`this.questions` would be undefined before init. so it willl throw error  "can not access undefined.getStateData(); "

Reply:
Then your pre-check can be of the form "expect ... toThrowError()".

Reply:
done!